components/browser/search/src/main/java/mozilla/components/browser/search/suggestions/SearchSuggestionClient.kt

```diff
@@ -0,0 +1,38 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.search.suggestions
+
+import mozilla.components.browser.search.SearchEngine
+
+/**
+ * Async function responsible for taking a URL and returning the results
+ */
+typealias SearchSuggestionFetcher = suspend (url: String) -> String?
+
+/**
+ *  Provides an interface to get search suggestions from a given SearchEngine.
+ */
+class SearchSuggestionClient private constructor(
+    private val searchEngine: SearchEngine,
+    private val fetcher: SearchSuggestionFetcher
+) {
+    companion object {
+        fun create(searchEngine: SearchEngine, fetcher: SearchSuggestionFetcher): SearchSuggestionClient? {
+            if (!searchEngine.canProvideSearchSuggestions) { return null }
+
+            return SearchSuggestionClient(searchEngine, fetcher)
+        }
+    }
+
+    /**
+     * Gets search suggestions for a given query
+     */
+    suspend fun getSuggestions(query: String): List<String>? {
+        val suggestionsURL = searchEngine.buildSuggestionsURL(query) ?: return null
+        val parser = selectResponseParser(searchEngine)
+
+        return fetcher(suggestionsURL)?.let { parser(it) }
+    }
```

Comment:
Regarding `JSONException` (and I assume `IOException`): I think we should either wrap them in a custom exception and make it explicit that this code throws it [(throws annotation)](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.jvm/-throws/index.html) _or_ just fail gracefully and return an empty list. I think for apps showing suggestions returning an empty list is probably the easiest to handle.